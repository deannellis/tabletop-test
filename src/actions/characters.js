import uuid from 'uuid';

// ADD_CHARACTER
export const addCharacter = (
    {
        abilities = {
            strength: 0,
            intelligence: 0,
            wisdom: 0,
            dexterity: 0,
            constitution: 0,
            charisma: 0
        }, 
        race = '', 
        charClass = '', 
        hp = 0, 
        xp = 0,
        name = '',
        description = ''
    } = {}
) => ({
    type: 'ADD_CHARACTER',
    character: {
        id: uuid(),
        abilities,
        race,
        charClass,
        hp,
        xp,
        name,
        description
    }
});

// REMOVE_CHARACTER

export const removeCharacter = ({ id } = {}) => ({
    type: 'REMOVE_CHARACTER',
    id
});

// EDIT_CHARACTER
export const editCharacter = (id, updates) => ({
    type: 'EDIT_CHARACTER',
    id,
    updates
});

