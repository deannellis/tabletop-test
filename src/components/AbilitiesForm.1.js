import React from 'react';

export default class AbilitiesForm extends React.Component {
    state = {
        strength: undefined,
        intelligence: undefined,
        wisdom: undefined,
        dexterity: undefined,
        constitution: undefined,
        charisma: undefined
    };
    
    renderButtons = () => {
        Object.keys(this.state).map((key) => {
            return 
            <div>
                <button type="button" onClick={this.onRollStrength} disabled={typeof this.state.strength != 'undefined'}>Strength</button>
                <p>Strength: {typeof this.state.strength == 'undefined' ? 'Roll to determine Strength' : this.state.strength }</p>
            
        });
    };
    
    onRollStrength = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({strength: rollValue}));
    };
    
    onRollIntelligence = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({intelligence: rollValue}));
    };
    
    onRollWisdom = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({wisdom: rollValue}));
    };
    
    onRollDexterity = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({dexterity: rollValue}));
    };
    
    onRollConstitution = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({constitution: rollValue}));
    };
    
    render() {
        return (
            <div>
                <form>
                    {this.renderButtons()}
                </form>
            </div>
        );
    }
}