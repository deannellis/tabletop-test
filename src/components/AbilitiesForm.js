import React from 'react';
import SingleRollDie from './SingleRollDie';
import uuid from 'uuid';

export default class AbilitiesForm extends React.Component {
    constructor(props){
        super(props);
        
        this.state = {
            strength: [],
            intelligence: [],
            wisdom: [],
            dexterity: [],
            constitution: [],
            charisma: [],
            error: '',
            refreshKey: 0,
            lastDieRolled: undefined,
            lastResult: undefined
        };
    }
    
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit} >
                    {this.renderAttribute('Strength', 'strength')}
                    {this.renderAttribute('Intelligence', 'intelligence')}
                    {this.renderAttribute('Wisdom', 'wisdom')}
                    {this.renderAttribute('Dexterity', 'dexterity')}
                    {this.renderAttribute('Constitution', 'constitution')}
                    {this.renderAttribute('Charisma', 'charisma')}
                    <button>Next Step</button>
                </form>
            </div>
        );
    }
    
    handleRollDie = (result, attributeKey) => {
        this.setState(() => ({
            lastResult: result,
            [attributeKey]: [...this.state[attributeKey], result]
        }));
    };
    
    onSubmit = (e) => {
        e.preventDefault();
        
        const sum = (accumulator, currentValue) => accumulator + currentValue;
        
        if(this.state.strength.length < 3 || this.state.intelligence.length < 3 || this.state.wisdom.length < 3 || this.state.dexterity.length < 3 || this.state.constitution.length < 3 || this.state.charisma.length < 3) {
            this.setState(() => ({
                error: 'Please roll all scores'
            }));
        } else {
            this.setState(() => ({
                error: ''
            }));
            this.props.onSubmit({
                id: uuid(),
                abilities: {
                    strength: this.state.strength.reduce(sum),
                    intelligence: this.state.intelligence.reduce(sum),
                    wisdom: this.state.wisdom.reduce(sum),
                    dexterity: this.state.dexterity.reduce(sum),
                    constitution: this.state.constitution.reduce(sum),
                    charisma: this.state.charisma.reduce(sum)
                },
                name: 'In-Progress Character',
                charClass: 'unkown'
            });
        } 
    }
    
    renderAttribute = (label, attributeKey) => {
        const attributeValue = this.state[attributeKey];
        const isDefined = attributeValue.length !== 0;
        let rollsString = '';
        if(this.state[attributeKey].length < 3){
            this.state[attributeKey].forEach(
                (roll) => {rollsString += `${roll} + `}
            );
        } else {
            let total = 0;
            this.state[attributeKey].forEach(
                (roll) => {total += roll}
            );
            rollsString = `${this.state[attributeKey][0]} + ${this.state[attributeKey][1]} + ${this.state[attributeKey][2]} = ${total}`;
        }
        
        const message = label + ': ' + (isDefined ? rollsString : 'roll to determine');
        return(
            <div>
                <p>{message}</p>
                <div className="dice-container">
                    <SingleRollDie dieType={6} handleRollDie={this.handleRollDie} attributeKey={attributeKey} updateRefreshKey={this.updateRefreshKey} refreshKey={this.state.refreshKey}/>
                    <SingleRollDie dieType={6} handleRollDie={this.handleRollDie} attributeKey={attributeKey} updateRefreshKey={this.updateRefreshKey} refreshKey={this.state.refreshKey}/>
                    <SingleRollDie dieType={6} handleRollDie={this.handleRollDie} attributeKey={attributeKey} updateRefreshKey={this.updateRefreshKey} refreshKey={this.state.refreshKey}/>
                </div>
            </div>
        );
    }
    
    updateRefreshKey = (newKey) => {
        this.setState(() => ({
            refreshKey: newKey
        }));
    };
}