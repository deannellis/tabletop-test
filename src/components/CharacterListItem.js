import React from 'react';
import { Link } from 'react-router-dom';

const CharacterListItem = ({ name, charClass, race, id }) => (
    <div>
        <h2>{name}</h2>
        <p>{race} - {charClass}</p>
        <Link to={`/view-char/${id}`}>View Character</Link>
    </div>    
);

export default CharacterListItem;
