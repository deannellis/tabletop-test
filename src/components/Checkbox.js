import React, { Component, PropTypes } from 'react';

class Checkbox extends React.Component {
    state = {
        isChecked: false,
    }
    
    toggleCheckboxChange = () => {
        const { handleCheckboxChange } = this.props;
        const { isChecked } = this.state;
    
        this.setState(({ isChecked }) => (
            {
                isChecked: !isChecked,
            }
        ));
    
        handleCheckboxChange(!isChecked);
    }
    
      render() {
        const { label } = this.props;
        const { isChecked } = this.state;
        let isDisabled;
        if(isChecked) {
            isDisabled = false;
        } else {
            isDisabled = this.props.isDisabled
        }
    
        return (
          <div className="checkbox">
            <label>
                <input
                    type="checkbox"
                    value={isChecked}
                    checked={isChecked}
                    onChange={this.toggleCheckboxChange}
                    disabled={isDisabled}
                />
    
              {label}
            </label>
          </div>
        );
      }
}

export default Checkbox;