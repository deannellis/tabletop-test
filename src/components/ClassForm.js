import React from 'react';
import { connect } from 'react-redux';
import spells from '../objects/spells';
import SpellListItem from './SpellListItem';

class ClassForm extends React.Component {
    constructor(props){
        super(props);
        
        this.state = {
            selectedOption: '',
            error: ''
        };
    }
    
    render() {
        const characters = this.props.characters;
        const lastCharacter = characters[characters.length - 1];
        let canCleric = true;
        let canFighter = true;
        let canMagicUser = true;
        let canThief = true;
        
        if(lastCharacter.race == 'dwarf') {
            canMagicUser = false;
        } else if (lastCharacter.race == 'halfling') {
            canFighter = false;
        } 
        
        if(lastCharacter.abilities.wisdom < 9) {
            canCleric = false;
        }
        
        if(lastCharacter.abilities.strength < 9) {
            canFighter = false;
        }
        
        if(lastCharacter.abilities.intelligence < 9) {
            canMagicUser = false;
        }
        
        if(lastCharacter.abilities.dexterity < 9) {
            canThief = false;
        }
        
        return (
            <div>
                <h2>Select Your Class</h2>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit} >
                    {canCleric 
                        ? 
                        <div>
                            <label>
                                <input type="radio" name="class" value="cleric"
                                    checked={this.state.selectedOption === 'cleric'}
                                    onChange={this.handleOptionChange}
                                />
                                Cleric
                            </label>
                        </div> 
                        : 
                        <p>Cleric: Your ability scores are insufficient</p>
                    }
                    {canFighter 
                        ? 
                        <div>
                            <label>
                                <input type="radio" name="class" value="fighter"
                                    checked={this.state.selectedOption === 'fighter'}
                                    onChange={this.handleOptionChange}
                                />
                                Fighter
                            </label>
                        </div> 
                        : 
                        <p>Fighter: Your ability scores are insufficient</p>
                    }
                    {canMagicUser 
                        ? 
                        <div>
                            <label>
                                <input type="radio" name="class" value="magic-user"
                                    checked={this.state.selectedOption === 'magic-user'}
                                    onChange={this.handleOptionChange}
                                />
                                Magic-User
                            </label>
                        </div> 
                        : 
                        <p>Magic-User: Your ability scores are insufficient</p>
                    }
                    {canThief 
                        ? 
                        <div>
                            <label>
                                <input type="radio" name="class" value="thief"
                                    checked={this.state.selectedOption === 'thief'}
                                    onChange={this.handleOptionChange}
                                />
                                Thief
                            </label>
                        </div> 
                        : 
                        <p>Thief: Your ability scores are insufficient</p>
                    }
                    <button>Next Step</button>
                </form>
            </div>
        )
    }
    
    handleOptionChange = (e) => {
        this.setState({
            selectedOption: e.target.value
        });
    }
    
    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.selectedOption) {
            this.setState(() => ({
                error: 'Please select a class'
            }));
        } else {
            const characters = this.props.characters;
            const lastCharacter = characters[characters.length - 1];
            this.setState(() => ({
                error: ''
            }));
            this.props.onSubmit( lastCharacter.id, {
                charClass: this.state.selectedOption
            });
        }
    }
}

const mapStateToProps = (state) => {
    return {
        characters: state.characters
    };
};

export default connect(mapStateToProps)(ClassForm);