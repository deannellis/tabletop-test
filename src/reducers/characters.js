//  CHARACTERS REDUCER
import actionIds from 'actions/actionIds';
const {ADD_CHARACTER, REMOVE_CHARACTER, EDIT_CHARACTER, COMPLETE_CHARACTER} = actionIds;
const charactersReducerDefaultState = [];

export default ( state = charactersReducerDefaultState, action ) => {
    switch (action.type) {
        case ADD_CHARACTER:
            return [
                ...state,
                action.character
            ];
        case REMOVE_CHARACTER:
            return state.filter(({ id }) => {
                return id !== action.id;
            });
        case EDIT_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    return {
                        ...character,
                        ...action.updates
                    }
                } else {
                    return character;
                }
            });
        case COMPLETE_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    let {inProgressStep, ...rest} = character;
                    console.log('rest'+rest);
                    return rest;
                } else {
                    return character;
                }
            });
        default:
            return state;
    }
};