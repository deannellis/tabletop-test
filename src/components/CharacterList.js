import React from 'react';
import { connect } from 'react-redux';
import CharacterListItem from './CharacterListItem';

const CharacterList = (props) => (
    <div>
        <h1>Character List</h1>
        {props.characters.length > 0 ?
            <div>
                {props.characters.map((character) => {
                    return <CharacterListItem key={character.id} {...character} />
                })}
            </div>
        :
            <p>You have no characters, Click below to create a character</p>
        }
    </div>    
);

const mapStateToProps = (state) => {
    return {
        characters: state.characters
    };
};

export default connect(mapStateToProps)(CharacterList);
