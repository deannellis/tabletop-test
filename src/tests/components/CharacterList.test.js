/* global expect */
import React from 'react';
import { shallow } from 'enzyme';
import { CharacterList } from '../../components/CharacterList';
import characters from '../fixtures/characters';

test('should render CharacterList with characters', () => {
    const wrapper = shallow(<CharacterList characters={characters} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render CharacterList with no characters message', () => {
    const wrapper = shallow(<CharacterList characters={[]} />);
    expect(wrapper).toMatchSnapshot();
});