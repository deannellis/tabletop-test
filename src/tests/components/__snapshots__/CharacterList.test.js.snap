// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CharacterList with characters 1`] = `
<div>
  <div>
    <CharacterListItem
      abilities={
        Object {
          "charisma": 9,
          "constitution": 9,
          "dexterity": 9,
          "intelligence": 14,
          "strength": 11,
          "wisdom": 11,
        }
      }
      charClass="magic-user"
      description="Loves his Mom"
      equipment={
        Array [
          0,
          1,
        ]
      }
      gold={109}
      hp={1}
      id="3a7b71ef-fb4e-4bde-a2c8-17cd5b9f4097"
      inProgressStep={2}
      key="3a7b71ef-fb4e-4bde-a2c8-17cd5b9f4097"
      name="Rick Slick"
      race="human"
      spells={
        Array [
          "1",
        ]
      }
      weapons={
        Array [
          "1",
        ]
      }
      xp={0}
    />
    <CharacterListItem
      abilities={
        Object {
          "charisma": 10,
          "constitution": 9,
          "dexterity": 9,
          "intelligence": 9,
          "strength": 10,
          "wisdom": 10,
        }
      }
      charClass="cleric"
      description="He has got goats running wild!"
      equipment={
        Array [
          1,
        ]
      }
      gold={123}
      hp={2}
      id="1d871f50-fd02-467a-8611-0599cde106f3"
      inProgressStep={2}
      key="1d871f50-fd02-467a-8611-0599cde106f3"
      name="Sweeney Dixon"
      race="dwarf"
      xp={0}
    />
    <CharacterListItem
      abilities={
        Object {
          "charisma": 9,
          "constitution": 14,
          "dexterity": 15,
          "intelligence": 10,
          "strength": 11,
          "wisdom": 9,
        }
      }
      charClass="fighter"
      description="publishes Golf Wolf Magazine"
      equipment={
        Array [
          0,
          3,
        ]
      }
      gold={41}
      hp={2}
      id="efc641b5-7643-416a-a627-ca269952a82a"
      inProgressStep={2}
      key="efc641b5-7643-416a-a627-ca269952a82a"
      name="Buck Pucker"
      race="halfling"
      xp={0}
    />
    <CharacterListItem
      abilities={
        Object {
          "charisma": 9,
          "constitution": 4,
          "dexterity": 9,
          "intelligence": 9,
          "strength": 9,
          "wisdom": 4,
        }
      }
      description="low wisdom character"
      equipment={
        Array [
          0,
          3,
        ]
      }
      gold={41}
      hp={2}
      id="c7562afd-cb8c-4166-9229-ead45db9b042"
      inProgressStep={2}
      key="c7562afd-cb8c-4166-9229-ead45db9b042"
      name="Wislow"
      race="human"
      xp={0}
    />
    <CharacterListItem
      abilities={
        Object {
          "charisma": 9,
          "constitution": 9,
          "dexterity": 9,
          "intelligence": 9,
          "strength": 4,
          "wisdom": 9,
        }
      }
      description="low strength character"
      equipment={
        Array [
          0,
          3,
        ]
      }
      gold={41}
      hp={2}
      id="73776b72-ddcd-421b-be9b-a2be60971f8b"
      inProgressStep={2}
      key="73776b72-ddcd-421b-be9b-a2be60971f8b"
      name="Strlow"
      race="human"
      xp={0}
    />
    <CharacterListItem
      abilities={
        Object {
          "charisma": 9,
          "constitution": 9,
          "dexterity": 9,
          "intelligence": 4,
          "strength": 9,
          "wisdom": 9,
        }
      }
      description="low intelligence character"
      equipment={
        Array [
          0,
          3,
        ]
      }
      gold={41}
      hp={2}
      id="f223009c-6247-4b44-a0fd-5be03ea4ca9e"
      inProgressStep={2}
      key="f223009c-6247-4b44-a0fd-5be03ea4ca9e"
      name="Intlow"
      race="human"
      xp={0}
    />
    <CharacterListItem
      abilities={
        Object {
          "charisma": 9,
          "constitution": 9,
          "dexterity": 4,
          "intelligence": 9,
          "strength": 9,
          "wisdom": 9,
        }
      }
      description="low dexterity character"
      equipment={
        Array [
          0,
          3,
        ]
      }
      gold={41}
      hp={2}
      id="8ce32d3f-b5a4-4d63-8188-6f7a38abe723"
      inProgressStep={2}
      key="8ce32d3f-b5a4-4d63-8188-6f7a38abe723"
      name="Dexlow"
      race="human"
      xp={0}
    />
  </div>
</div>
`;

exports[`should render CharacterList with no characters message 1`] = `
<div>
  <p>
    You have no characters, Click below to create a character
  </p>
</div>
`;
